src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs

```diff
@@ -25,25 +27,66 @@ public override bool Execute()
         {
             ns = "urn:schemas-microsoft-com:asm.v1";
             XElement bindingRedirectAssemblies = new XElement(ns + "assemblyBinding");
-            foreach (ITaskItem assembly in AssemblyIdentities)
+            foreach (ITaskItem assembly in Assemblies)
             {
-                string publicKeyToken = assembly.GetMetadata("PublicKeyToken");
-                string assemblyVersion = assembly.GetMetadata("Version");
-                string assemblyName = assembly.GetMetadata("Name");
+                AssemblyName result = null;
+                try
+                {
+                    using (var assemblyStream = new FileStream(assembly.ItemSpec, FileMode.Open, FileAccess.Read, FileShare.Delete | FileShare.Read))
+                    using (PEReader peReader = new PEReader(assemblyStream, PEStreamOptions.LeaveOpen))
+                    {
+                        if (peReader.HasMetadata)
+                        {
+                            MetadataReader reader = peReader.GetMetadataReader();
+                            if (reader.IsAssembly)
+                            {
+                                var assemblyDef = reader.GetAssemblyDefinition();
+
+                                result = new AssemblyName();
+                                result.Name = reader.GetString(assemblyDef.Name);
+                                result.CultureName = reader.GetString(assemblyDef.Culture);
+                                result.Version = assemblyDef.Version;
+
+                                if (!assemblyDef.PublicKey.IsNil)
+                                {
+                                    result.SetPublicKey(reader.GetBlobBytes(assemblyDef.PublicKey));
+                                }
+                            }
+                        }
+                        else
+                        {
+                            //Native
+                            continue;
+                        }
+                    }
+                }
+                catch (BadImageFormatException)
+                {
+                    // not a PE
+                    continue;
+                }
+
+                string publicKeyToken = string.Empty;
+                if (result.GetPublicKeyToken() != null)
+                { 
+                    byte[] pt = result.GetPublicKeyToken();
```

Comment:
Rather than call this method multiple times, would it be preferable to store it in a variable before it's used in the if condition on line 70?

This comment also applies to the GetMetadata calls on lines 80-82.